backbone.js

```diff
@@ -1621,11 +1620,12 @@
       // Normalize the fragment.
       fragment = this.getFragment(fragment || '');
 
-      // Don't include a trailing slash on the root.
+      // Append a trailing slash if needed.
       var root = this.root;
-      if (fragment === '' || fragment.charAt(0) === '?') {
-        root = root.slice(0, -1) || '/';
+      if (fragment && this._usePushState) {
+        if (root.charAt(root.length - 1) !== '/' && fragment.charAt(0) !== '?') root += '/';
       }
```

Comment:
I would still need to check `fragment !== ''` and `root.charAt(root.length - 1) !== '/'` to avoid adding an unnecessary slash.

```javascript
Backbone.history.start({ root: '/root' });
// Shouldn't have a trailing slash
Backbone.history.navigate('');
location.pathname === '/root';

Backbone.history.start({ root: '/root/' });
// Shouldn't have a double slash
Backbone.history.navigate('test');
location.pathname === '/root/test';